src/middlewared/middlewared/plugins/virt/instance.py

```diff
@@ -0,0 +1,265 @@
+import aiohttp
+
+from middlewared.service import (
+    CRUDService, ValidationErrors, filterable, job, private
+)
+from middlewared.utils import filter_list
+
+from middlewared.api import api_method
+from middlewared.api.current import (
+    VirtInstanceEntry,
+    VirtInstanceCreateArgs, VirtInstanceCreateResult,
+    VirtInstanceUpdateArgs, VirtInstanceUpdateResult,
+    VirtInstanceDeleteArgs, VirtInstanceDeleteResult,
+    VirtInstanceStartArgs, VirtInstanceStartResult,
+    VirtInstanceStopArgs, VirtInstanceStopResult,
+    VirtInstanceRestartArgs, VirtInstanceRestartResult,
+    VirtInstanceImageChoicesArgs, VirtInstanceImageChoicesResult,
+)
+from .utils import incus_call, incus_call_and_wait
+
+
+LC_IMAGES_SERVER = 'https://images.linuxcontainers.org'
+LC_IMAGES_JSON = f'{LC_IMAGES_SERVER}/streams/v1/images.json'
+
+
+class VirtInstanceService(CRUDService):
+
+    class Config:
+        namespace = 'virt.instance'
+        cli_namespace = 'virt.instance'
+        entry = VirtInstanceEntry
+        role_prefix = 'VIRT_INSTANCE'
+
+    @filterable
+    async def query(self, filters, options):
+        """
+        Query all instances with `query-filters` and `query-options`.
+        """
+        results = (await incus_call('1.0/instances?filter=&recursion=2', 'get'))['metadata']
```

Comment:
This fails when incus is unset because incus socket is not running all the time and we get a traceback (similar to the comment where we should have a safety net of sorts to prevent this from happening and it should return an empty list)